Convert testing framework from nose to pytest

plaplant opened the pull request:
This PR converts the testing framework of hera_cal from nose (which is deprecated and may go away in the future) to pytest, which is more fully featured and has much broader support. The docs are [here](https://docs.pytest.org/en/latest/contents.html), but the tl;dr is:
- Tests use bare `assert` statements rather than `nt.assert_true` or `nt.assert_equal`-type statements.
- Raising of errors is testing by calling `pytest.raises`, with similar syntax to nose's `nt.assert_raises` or unittest's `self.assertRaises`.
- Specific warnings (such as floating point exceptions) can be easily filtered using function decorators (some examples of which are [here](https://github.com/HERA-Team/hera_cal/blob/940f882ca9e15ccc5d45fe93e3285db3444b8534/hera_cal/tests/test_abscal.py#L31) and [here](https://github.com/HERA-Team/hera_cal/blob/940f882ca9e15ccc5d45fe93e3285db3444b8534/hera_cal/tests/test_delay_filter.py#L22)).
- Tests can be invoked from anywhere inside of the repo, by calling `pytest` to run all tests or, e.g., `pytest test_delay_filter.py::Test_DelayFilter::test_run_filter`.

The PR updates the travis script accordingly, and also fixes a few small warnings that were being hidden behind the large number of (relatively) benign warnings triggered by the tests.

@jsdillon there isn't a huge time pressure on this, so feel free to wait until after IDR2.2 is done before reviewing/merging. Just wanted to have your eyes on it so we can make sure new tests use this framework going forward.

jsdillon commented on hera_cal/tests/test_redcal.py:
```diff
@@ -414,14 +414,14 @@ def test_solver(self):
-            np.testing.assert_equal(data['g_0_Jxx * g_1_Jxx_ * u_0_xx'], d[0, 1, 'xx'])
-            np.testing.assert_equal(data['g_1_Jxx * g_2_Jxx_ * u_0_xx'], d[1, 2, 'xx'])
-            np.testing.assert_equal(data['g_0_Jxx * g_2_Jxx_ * u_1_xx'], d[0, 2, 'xx'])
+            assert np.allclose(data['g_0_Jxx * g_1_Jxx_ * u_0_xx'], d[0, 1, 'xx'])
```
Same problem.

jsdillon commented on hera_cal/tests/test_redcal.py:
```diff
@@ -414,14 +414,14 @@ def test_solver(self):
-            np.testing.assert_equal(wgts['g_0_Jxx * g_1_Jxx_ * u_0_xx'], w[0, 1, 'xx'])
-            np.testing.assert_equal(wgts['g_1_Jxx * g_2_Jxx_ * u_0_xx'], w[1, 2, 'xx'])
-            np.testing.assert_equal(wgts['g_0_Jxx * g_2_Jxx_ * u_1_xx'], w[0, 2, 'xx'])
+            assert np.allclose(wgts['g_0_Jxx * g_1_Jxx_ * u_0_xx'], w[0, 1, 'xx'])
```
And again.

jsdillon commented on hera_cal/tests/test_redcal.py:
```diff
@@ -511,9 +511,9 @@ def test_logcal(self):
         om.make_sol_finite(sol)
         for red in reds:
-            np.testing.assert_array_equal(sol[red[0]], 0.0)
+            assert np.allclose(sol[red[0]], 0.0)
```
And here.

jsdillon commented on hera_cal/tests/test_smooth_cal.py:
```diff
@@ -55,24 +54,24 @@ def test_freq_filter(self):
         wgts[0, 0:8] = 0
         freqs = np.linspace(100., 200., 10, endpoint=False) * 1e6
         ff, info = smooth_cal.freq_filter(gains, wgts, freqs, skip_wgt=.5)
-        np.testing.assert_array_equal(ff[0, :], gains[0, :])
```
Here again.

jsdillon commented on hera_cal/tests/test_smooth_cal.py:
```diff
@@ -121,111 +120,112 @@ def test_pick_reference_antenna(self):
+        assert (54, 'Jxx') in self.cs.flag_grids
+        assert self.cs.gain_grids[54, 'Jxx'].shape == (180, 1024)
+        assert self.cs.flag_grids[54, 'Jxx'].shape == (180, 1024)
+        assert np.allclose(self.cs.flag_grids[54, 'Jxx'][60:120, :], True)
```
Here again.

jsdillon commented on hera_cal/tests/test_utils.py:
```diff
@@ -323,73 +311,73 @@ def test_lst_rephase():
+    assert nObs.shape == (5, 10)
+    assert chisq.dtype == float
+    assert nObs.dtype == int
+    assert np.allclose(chisq, 4.0)
```
Same comment

jsdillon commented on hera_cal/tests/test_utils.py:
```diff
@@ -323,73 +311,73 @@ def test_lst_rephase():
-    np.testing.assert_array_equal(chisq, 2.0)
-    np.testing.assert_array_equal(nObs, 2)
-    nt.assert_false((1, 2, 'xx') in model)
+    assert np.allclose(chisq, 2.0)
```
And here.

jsdillon commented on hera_cal/tests/test_utils.py:
```diff
@@ -398,12 +386,12 @@ def test_chisq():
-    np.testing.assert_array_equal(chisq_per_ant[1, 'Jxx'], 2.0)
-    np.testing.assert_array_equal(nObs_per_ant[0, 'Jxx'], 2)
-    np.testing.assert_array_equal(nObs_per_ant[1, 'Jxx'], 2)
+    assert np.allclose(chisq, 2.0)
```
And here.

jsdillon commented on hera_cal/tests/test_utils.py:
```diff
@@ -412,57 +400,54 @@ def test_chisq():
+    assert 'Jxx' in nObs
+    assert chisq['Jxx'].shape == (5, 10)
+    assert nObs['Jxx'].shape == (5, 10)
+    assert np.allclose(chisq['Jxx'], 1.0)
```
And here.

jsdillon commented on hera_cal/tests/test_vis_clean.py:
```diff
@@ -113,50 +116,50 @@ def test_vis_clean(self):
         V.vis_clean(keys=[(24, 25, 'xx'), (24, 24, 'xx')], ax='both', max_frate=10., overwrite=True,
                     filt2d_mode='plus')
-        nt.assert_true('success' in V.clean_info[(24, 25, 'xx')])
+        'success' in V.clean_info[(24, 25, 'xx')]
```
Are you missing an assert here?

plaplant replied:
Good catch, thanks!

plaplant commented:
Thanks for the detailed review @jsdillon. Just to respond to the two high level points.

1. I used the `np.isclose` construction for boolean values because the "right" way to test for booleans (asserting that `var is True`) fails, because the values are the `np.bool` datatype and not the singleton `True`, and so that assert fails. The `np.isclose` function accepts all numpy datatypes, and in particular will return `False` if two boolean values are different. I agree that it's odd to see it phrased that way, but it Does The Right Thing. Though as you suggest, we could switch to using `np.all`/`np.any` to test for boolean equality.
2. I don't personally use the `np.testing` module, and didn't realize it had those benefits. It should be possible to switch them back, as pytest is generally pretty flexible and can mix and match with different frameworks. In my mind, one benefit of `allclose` is requiring the user to explicitly condone NaNs, which by default are not equal (but are considered equal in `numpy.testing`). I'd _really_ like to see us get to a point where we avoid producing them in the first place. But I can switch back if that's really a showstopper.

jsdillon commented:
On 1) I think we should switch them to np.all / np.any, if only just for readability. When I see `close`, I immediately think I'm looking at a floating point comparison.

On 2) I'm kinda torn. I'm fine with any changes that are not `array` comparisons. I'd prefer to leave it when they are array comparisons. I don't feel very strongly, however, so I'll leave it up to you to decide how much time you want to spend.

plaplant commented:
@jsdillon I've gone back and addressed the main concerns above. On 1), I replaced all the `isclose` comparisons for boolean types. On 2), I've reverted back to `np.testing` calls (where they weren't generating warnings). I'll also respond to smaller inline comments.

plaplant commented on hera_cal/vis_clean.py:
```diff
@@ -825,6 +825,7 @@ def zeropad_array(data, binvals=None, zeropad=0, axis=-1, undo=False):
             if undo:
                 s = [slice(None) for j in range(data.ndim)]
                 s[ax] = slice(zeropad[i], -zeropad[i])
+                s = tuple(s)
```
If we don't do this, a warning is generated:
```
Using a non-tuple sequence for multidimensional indexing is
deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the
future this will be interpreted as an array index, `arr[np.array(seq)]`,
which will result either in an error or a different result.
```
I can't do a tuple comprehension, because tuples are immutable, and so the elements can't be changed after it's created (as in the line immediately prior).
